keps/sig-storage/3751-volume-attributes-class/README.md

```diff
@@ -0,0 +1,1279 @@
+  // State set when modify volume controller starts modifying the volume in control-plane
+  PersistentVolumeClaimControllerModifyVolumeInProgress PersistentVolumeClaimModifyVolumeStatus = "ControllerModifyVolumeInProgress"
+  // State set when modify volume has failed in modify volume controller with a terminal error.
+  // Transient errors such as timeout should not set this status and should leave ModifyVolumeStatus
+  // unmodified, so as modify volume controller can resume the volume modification.
+  PersistentVolumeClaimControllerModifyVolumeFailed PersistentVolumeClaimModifyVolumeStatus = "ControllerModifyVolumeFailed"
+)
+
+```
+
+
+#### Quota
+
+Cluster administrators can use K8s quota to specify how many PVCs can use a specific VolumeAttributesClass. There is no fine grained parameter specific quota control for this feature. Fine grained attributes key-value pair quota is not in scope of this KEP.
+
+### CSI API
+The CSI create request will be extended to add mutable parameters. A new ControllerModifyVolume RPC will be added.
+
+```
+// ControllerServer is the server API for Controller service.
+type ControllerServer interface {
+        ...
+    rpc ControllerModifyVolume (ModifyVolumeRequest)
+￼        returns (ModifyVolumeResponse) {
+￼            option (alpha_method) = true;
+￼        }
+        ...
+}
+
+message ModifyVolumeRequest {
+  // Contains identity information for the existing volume.
+  // This field is REQUIRED.
+  string volume_id = 1
+  // This field is OPTIONAL.This allows the CO to specify the
+  // mutable parameters to apply. 
+  map<string, string> mutable_parameters = 2;
+}
+message ModifyVolumeResponse {}
+
+message CreateVolumeRequest {
+  ...
+  // See CreateVolumeRequest.parameters.
+  // This field is OPTIONAL.
+  map<string, string> parameters = 4;
+  // This field is OPTIONAL. This allows the CO to specify the
+  // volume attributes class parameters to apply. 
+  map<string, string> mutable_parameters = 8;
+}
+```
+
+### User Stories (Optional)
+
+#### Default VolumeAttributesClass
+
+A default VolumeAttributesClass can be specified for the Kubernetes cluster. This default VolumeAttributesClass is then used to dynamically provision storage for PersistentVolumeClaims that do not require any specific VolumeAttributesClass. A cluster admin can use annotation to manage default VolumeAttributesClass. The default VolumeAttributesClass has an annotation volumeattributesclass.kubernetes.io/is-default-class set to true. Any other value or absence of the annotation is interpreted as false. 
+
+#### Pre-provisioned Volume
+
+The cluster admin has created a group of PVs with VolumeAttributesClass and an end user is binding PVCs to pre-provisioned PVs. The PVCs will honor the VolumeAttributesClass. It is up to the cluster admin to make sure the parameters in the PVs matching the parameters in the VolumeAttributesClass. This is similar to StorageClass.
```

Comment:
Do we need to add finalizers based on unbound PVs referencing a VAC?

Reply:
We can. But does it need to be a P0 feature?